### PR TITLE
Add cron trigger support

### PR DESCRIPTION
**Pull request by bsaavedra17**
Added the ability to specify a cron teamcity build trigger schedule.
The configuration would look like the following:
```
resource "teamcity_build_trigger_schedule" "schedule" {
  build_config_id = module.service.build_configs["commit"].id

  rules = [""]
  cron_schedule {
    seconds      = "5"
    minutes      = "8"
    hours        = "21"
    day_of_week  = "*"
    day_of_month = "*"
    month        = "12"
  }
  schedule = "cron"
}
```
Also added validation for the cron expression. Tested by using out of range numbers for the fields, and saw the expected errors.

### CODE REVIEW

File: teamcity/resource_build_config.go [hunk 1]
```diff
@@ -826,11 +826,19 @@ func expandBuildFeature(raw interface{}) (api.BuildFeature, error) {
...
 	out := make([]api.Step, 0)
 	in := list.([]interface{})
+	names := make(map[string]struct{})
 	for _, i := range in {
 		s, err := expandBuildStep(i)
 		if err != nil {
 			return nil, err
 		}
+
+		if _, exist := names[s.GetName()]; exist {
```
Comment (atavakoliyext) [inline teamcity/resource_build_config.go]: This change seems unrelated to cron trigger support.

File: teamcity/resource_build_trigger_schedule.go [hunk 2]
```diff
@@ -260,6 +314,51 @@ func expandTriggerScheduleOptions(d *schema.ResourceData) (*api.TriggerScheduleO
...
+		cronSchedule.Month = v.(string)
+	}
+	if v, ok := raw["day_of_week"]; ok {
+		cronSchedule.DayOfWeek = v.(string)
+	}
+	if v, ok := raw["year"]; ok {
+		cronSchedule.Year = v.(string)
+	}
+	// Validate the cron expression
+	cronParser := cron.NewParser(cron.Second | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)
```
Comment (atavakoliyext) [inline teamcity/resource_build_trigger_schedule.go:346]: Just to confirm, this package has identical (or at least is a proper subset) of the configuration supported by TeamCity?

File: teamcity/resource_build_trigger_schedule.go [hunk 3]
```diff
@@ -65,6 +66,44 @@ func resourceBuildTriggerSchedule() *schema.Resource {
 				Elem:     &schema.Schema{Type: schema.TypeString},
 				Optional: true,
 			},
+			"cron_schedule": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Computed: true,
```
Comment (atavakoliyext) [inline teamcity/resource_build_trigger_schedule.go]: What's the reason for making this computed?

File: teamcity/resource_build_trigger_schedule.go [hunk 4]
```diff
@@ -65,6 +66,44 @@ func resourceBuildTriggerSchedule() *schema.Resource {
 				Elem:     &schema.Schema{Type: schema.TypeString},
 				Optional: true,
 			},
+			"cron_schedule": {
```
Comment (atavakoliyext) [inline teamcity/resource_build_trigger_schedule.go:69]: As with the other schedule-related parameters, should this also have `ForceNew: true`?

File: teamcity/resource_build_trigger_schedule.go [hunk 5]
```diff
@@ -65,6 +66,44 @@ func resourceBuildTriggerSchedule() *schema.Resource {
...
+				Type:     schema.TypeList,
```
Comment (atavakoliyext) [inline teamcity/resource_build_trigger_schedule.go:70]: Just to confirm, is this the typical way to add a struct-style configuration (single-element list)? Are there any similar examples of this pattern?

File: teamcity/resource_build_trigger_schedule.go [hunk 6]
```diff
@@ -65,6 +66,44 @@ func resourceBuildTriggerSchedule() *schema.Resource {
...
+				Optional: true,
+				MaxItems: 1,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"seconds": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (atavakoliyext) [inline teamcity/resource_build_trigger_schedule.go]: To confirm, if one of these is left out, is the behavior identical to having `*`?

File: teamcity/resource_build_trigger_schedule.go [hunk 7]
```diff
@@ -27,12 +28,12 @@ func resourceBuildTriggerSchedule() *schema.Resource {
 				Type:         schema.TypeString,
 				ForceNew:     true,
 				Required:     true,
-				ValidateFunc: validation.StringInSlice([]string{"daily", "weekly"}, false),
+				ValidateFunc: validation.StringInSlice([]string{"daily", "weekly", "cron"}, false),
```
Comment (atavakoliyext) [inline teamcity/resource_build_trigger_schedule.go:31]: Is it possible to enforce the presence of `cron_schedule` (and perhaps, absence of the hour/minute/... fields) if `schedule = "cron"`?

File: teamcity/resource_build_trigger_schedule.go (hunk 6 above)
Comment (bsaavedra17) [reply teamcity/resource_build_trigger_schedule.go]: At this point no, but I think `*` would be a good value to have as a default. Added this

File: teamcity/resource_build_trigger_schedule.go (hunk 4 above)
Comment (bsaavedra17) [reply teamcity/resource_build_trigger_schedule.go:69]: Yup, changed it to do just that.

File: teamcity/resource_build_trigger_schedule.go (hunk 7 above)
Comment (bsaavedra17) [reply teamcity/resource_build_trigger_schedule.go:31]: Each field has support for a ValidateFunc (more documentation [here](https://github.com/yext/terraform-provider-teamcity/blob/yext/vendor/github.com/hashicorp/terraform/helper/schema/schema.go#L243)). However, it doesn't look like it has support for dependencies on another field, or for fields of type schema.TypeList. I'll add logic to have the apply stage error out if `schedule = "cron"` is specified, and `cron_schedule` is not defined.

File: teamcity/resource_build_trigger_schedule.go (hunk 2 above)
Comment (bsaavedra17) [reply teamcity/resource_build_trigger_schedule.go:346]: From my investigation it does, at least for the parsing bit we care about here. Documentation [here](https://github.com/robfig/cron)

Comment (atavakoliyext) [reply teamcity/resource_build_trigger_schedule.go:346]: SGTM, thanks for confirming!

File: teamcity/resource_build_trigger_schedule.go (hunk 7 above)
Comment (atavakoliyext) [reply teamcity/resource_build_trigger_schedule.go:31]: SGTM!

File: teamcity/resource_build_trigger_schedule.go (hunk 6 above)
Comment (atavakoliyext) [reply teamcity/resource_build_trigger_schedule.go]: Another reasonable option IMO is to make every field required. This is also the case with cron itself and the explicitness could help in avoiding bugs related to having the wrong idea about default behavior (e.g. sometime assuming the default is 0).

Your call!

File: teamcity/resource_build_trigger_schedule.go (hunk 3 above)
Comment (bsaavedra17) [reply teamcity/resource_build_trigger_schedule.go]: Removed.

File: teamcity/resource_build_trigger_schedule.go (hunk 5 above)
Comment (bsaavedra17) [reply teamcity/resource_build_trigger_schedule.go:70]: I mostly based this structure off of the way `settings` in `resource_build_configs` are handled. Code [here](https://github.com/yext/terraform-provider-teamcity/blob/yext/teamcity/resource_build_config.go#L183). A scema.TypeMap didn't seem to work with this, since we wanted to have specific keys, and would then have to add validation for those keys.

Comment (atavakoliyext) [reply teamcity/resource_build_trigger_schedule.go:70]: Thanks for clarifying! LGTM

File: teamcity/resource_build_trigger_schedule.go (hunk 6 above)
Comment (bsaavedra17) [reply teamcity/resource_build_trigger_schedule.go]: I like this idea, having every field required makes a user have to think about all the fields, and unable to rely on any defaults. Done, it'll be included in the next patchset.

File: teamcity/resource_build_config.go (hunk 1 above)
Comment (bsaavedra17) [reply teamcity/resource_build_config.go]: I'm working on fixing that, not sure why they got included.

Comment (bsaavedra17) [reply teamcity/resource_build_config.go]: Done

**Review by atavakoliyext — APPROVED** [review]
Thanks @bsaavedra17!
